torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1152,6 +1158,76 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
   abort(abortReason);
 }
 
+void dump_debugging_info() {
+  LOG(ERROR) << dump_nccl_trace();
+}
+
+size_t hashHeartBeat(const HeartBeat& heartBeat) {
+  std::hash<size_t> id_hasher;
+  std::hash<std::uint8_t> type_hasher;
+  return id_hasher(heartBeat.pg_id_) ^ id_hasher(heartBeat.seq_id_) ^
+      type_hasher(static_cast<std::uint8_t>(heartBeat.op_type_));
+}
+
+void ProcessGroupNCCL::ncclCommsMonitor() {
+  size_t previousHashValue = 0;
+  int counter = -1;
+
+  while (true) {
+    std::unique_lock<std::mutex> lock(heartbeatMutex_);
```

Comment:
I am not sure if I need to also add try-catch here like watchdog thread does.

Reply:
iiuc this line will hang the watcher thread.  we don't want that

that's the reason that we discussed using pipes to communicate the heartbeat.  In the pipe model (or message Q model) there is no need to share a lock between the two threads.  This lets the 'watcher' thread simply loop/peek at the recv end of the pipe/queue and have no risk of hanging. 

See https://www.bogotobogo.com/cplusplus/multithreading_ipc.php -- you can compare the queue/pipe example with the shared variable example.

Reply:
wait, then when it calls `wait_for`, it will release the mutex lock right? Let me look into the pipe example. Also cc: @kwen2501

Reply:
That's why I thought having a heartbeat variable shared between threads but without a lock paired to it is better than with a lock paired to it. Eventually, as long as the monitoring thread can sense a change in that heartbeat variable, we are good. The read-after-write race is not a concern in this case.

Reply:
Oh, if the shared variable is a queue, then it is a whole different story. I was thinking of the shared variable being a `long` counter.

Reply:
I think we do need a queue here, so do we all agree that we can use a message queue instead? I prefer mq rather than a pipe. @kwen2501 @wconstab

Reply:
if we only care about 'step 1' of your watchdog plan, we could use a shared varaible without a lock.  the variable would be an int that gets incremented by one thread and watched by the other thread.  The 'watcher/other' thread would sleep a certain time, wake, check if the counter increased, if not, raise alarm.

but since we also care about 'step 2', we want a way to decouple the 'watcher' thread out of process. this was where the 'pipe' idea came from.  Now we are publishing a heartbeat message to the pipe from one thread, and seeing it from the other thread.